django/urls/resolvers.py

```diff
@@ -151,48 +139,218 @@ def _check_pattern_startswith_slash(self):
+        warnings = []
+        warnings.extend(self._check_pattern_startswith_slash())
+        if not self._is_endpoint:
+            warnings.extend(self._check_include_trailing_dollar())
+        return warnings
+
+    def _compile(self, regex):
+        """
+        Compile and return the given regular expression.
+        """
+        try:
+            return re.compile(regex)
+        except re.error as e:
+            raise ImproperlyConfigured(
+                '"%s" is not a valid regular expression: %s' % (regex, e)
+            )
+
+    def __str__(self):
+        return self._regex
+
+
+_PATH_PARAMETER_COMPONENT_RE = re.compile(
+    '<(?:(?P<converter>[^:]+):)?(?P<parameter>\w+)>'
+)
+
+
+def _route_to_regex(route, is_endpoint=False):
+    parts = ['^']
+    converters = {}
+    while True:
+        match = _PATH_PARAMETER_COMPONENT_RE.search(route)
+        if not match:
+            parts.append(re.escape(route))
+            break
+
+        parts.append(re.escape(route[:match.start()]))
+        route = route[match.end():]
+
+        parameter = match.group('parameter')
+        if not parameter.isidentifier():
+            raise ImproperlyConfigured(
+                "Parameter name {!r} is not a valid identifier.".format(parameter)
+            )
+        raw_converter = match.group('converter')
+        if raw_converter is None:
+            # If no converter is specified, the default is ``string``.
+            raw_converter = 'string'
+        converter = get_converter(raw_converter)
+        converters[parameter] = converter
+        parts.append('(?P<' + parameter + '>' + converter.regex + ')')
+    if is_endpoint:
+        parts.append('$')
+    return ''.join(parts), converters
+
+
+class RoutePattern(CheckURLMixin):
+    def __init__(self, route, name=None, is_endpoint=False):
+        self._route = route
+        self._regex_dict = {}
+        self._is_endpoint = is_endpoint
+        self.name = name
+        self.converters = _route_to_regex(str(route), is_endpoint)[1]
+
+    regex = LocaleRegexDescriptor('_route')
+
+    def match(self, path):
+        match = self.regex.search(path)
+        if match:
+            # If there are any named groups, use those as kwargs, ignoring
+            # non-named groups. Otherwise, pass all non-named arguments as
+            # positional arguments.
+            kwargs = match.groupdict()
+            args = () if kwargs else match.groups()
+            # In both cases, pass any extra_kwargs as **kwargs.
+
+            for key, value in kwargs.items():
+                if key in self.converters:
+                    converter = self.converters[key]
+                    try:
+                        kwargs[key] = converter.to_python(value)
+                    except ValueError:
+                        return None
+            return path[match.end():], args, kwargs
+        return None
+
+    def check(self):
+        return self._check_pattern_startswith_slash()
```

Comment:
I'm not sure if this check can fail. When writing something like `path('/foo', ...)`, the resulting pattern will be escaped to `^\/foo$`. Currently the method is untested.